File: crates/better-auth-rs-utils/src/password.node.ts
import { scrypt, randomBytes } from "node:crypto";

const config = {
	N: 16384,
	r: 16,
	p: 1,
	dkLen: 64,
};

function generateKey(password: string, salt: string): Promise<Buffer> {
	return new Promise((resolve, reject) => {
		scrypt(
			password.normalize("NFKC"),
			salt,
			config.dkLen,
			{
				N: config.N,
				r: config.r,
				p: config.p,
				maxmem: 128 * config.N * config.r * 2,
			},
			(err, key) => {
				if (err) reject(err);
				else resolve(key);
			},
		);
	});
}

export async function hashPassword(password: string): Promise<string> {
	const salt = randomBytes(16).toString("hex");
	const key = await generateKey(password, salt);
	return `${salt}:${key.toString("hex")}`;
}

export async function verifyPassword(
	hash: string,
	password: string,
): Promise<boolean> {
	const [salt, key] = hash.split(":");
	if (!salt || !key) {
		throw new Error("Invalid password hash");
	}
	const targetKey = await generateKey(password, salt);
	return targetKey.toString("hex") === key;
}

File: crates/better-auth-rs-utils/src/password_node.rs
//! Native-`node:crypto` scrypt password hashing (port of `password.node.ts`).
//!
//! `password.node.ts` exists upstream so Node can use the built-in `node:crypto.scrypt` rather than
//! `@noble/hashes` (`password.ts`). Both sides use **identical** parameters (`N=16384`, `r=16`,
//! `p=1`, `dkLen=64`), the same hex-string salt, NFKC normalization, and the `"{salt}:{key}"`
//! storage format, so their outputs are byte-for-byte interchangeable (upstream even asserts this
//! cross-compatibility in its tests).
//!
//! In Rust there is no Node-vs-noble split: a single audited `scrypt` implementation backs both.
//! This module therefore re-exports [`crate::password`], preserving the 1:1 file mapping with
//! `password.node.ts` while keeping a single source of truth.

pub use crate::password::{PasswordError, hash_password, verify_password};

#[cfg(test)]
#[path = "password_node.test.rs"]
mod password_node_tests;
